File: typescript/difference-of-squares/difference-of-squares.ts
function square(x: number) {
    return x * x
}

function sum(numbers: number[]) {
    return numbers.reduce((x, y) => x + y, 0)
}

function* range (begin: number, end: number) {
    for (let i = begin; i <= end; i++) {
        yield i
    }
}

export default class Squares {
    private n: number

    constructor(n: number) {
        this.n = n
    }

    get squareOfSum() {
        return square(sum(this.range))
    }

    get sumOfSquares() {
        return sum(this.range.map(square))
    }

    get difference() {
        return this.squareOfSum - this.sumOfSquares
    }

    private get range() {
        return [...range(1, this.n)]
    }
}

File: clojure/difference-of-squares/src/difference_of_squares.clj
(ns difference-of-squares)

(defn- square [n] (* n n))
(defn- natural-numbers [n] (range 1 (inc n)))

(defn sum-of-squares [n] (reduce + (map square (natural-numbers n))))
(defn square-of-sum [n] (square (reduce + (natural-numbers n))))
(defn difference [n] (- (square-of-sum n) (sum-of-squares n)))

